### PR TITLE
Optimize PenetrationThread::getSidesOnPrimaryBoundary

### LINKED ISSUE #16515: Optimize PenetrationThread::getSidesOnPrimaryBoundary
## Reason

Algorithm in `PenetrationThread::getSidesOnPrimaryBoundary` is `O(N^2)` and that slows MOOSE down for large meshes.

## Design

`_bc_tuples` does not need to be traversed because it is ordered. Use `std::equal_range` to speed up solve. 

## Impact

Performance improvement.

### PR DESCRIPTION
**Pull request by fdkong**
Closes #16515

### CODE REVIEW

**Comment by fdkong** [comment]
@snschune Could you try it out to check if this has any improvement?

Or could you tell me what example I should look at?

File: framework/src/geomsearch/PenetrationThread.C [hunk 1]
```diff
@@ -1764,8 +1764,21 @@ PenetrationThread::getSidesOnPrimaryBoundary(std::vector<unsigned int> & sides,
...
+    bool operator()(dof_id_type id, const libMesh::BoundaryInfo::BCTuple & tup) const
+    {
+      return id < std::get<0>(tup);
+    }
+  };
+
+  auto range = std::equal_range(_bc_tuples.begin(), _bc_tuples.end(), elem->id(), Comp{});
+
+  for (auto & t = range.first; t != range.second; ++t)
+    if (std::get<2>(*t) == static_cast<boundary_id_type>(_primary_boundary))
```
Comment (lindsayad) [inline framework/src/geomsearch/PenetrationThread.C:1782]: I don't know why this `static_cast` was there before. `BoundaryID` is just a `typedef` of `boundary_id_type`

**Review by lindsayad — APPROVED** [review]
It would be nice to get some report of performance improvement (this should indeed improve performance since we're going from O(N) to O(log(N) here) before merging.

File: framework/src/geomsearch/PenetrationThread.C (hunk 1 above)
Comment (snschune) [reply framework/src/geomsearch/PenetrationThread.C:1782]: I'll check performance right now.

**Comment by snschune** [comment]
I'll check performance right now.

**Comment by snschune** [comment]
About 1 million elements, hexahedral 3D. Before: 572.38 s, After: 10.63 s. I'd call that good.

**Review by snschune — APPROVED** [review]

**Comment by lindsayad** [comment]
I love that

**Pull request merged**